pkg/events/derive/symbols_collision.go

```diff
@@ -0,0 +1,299 @@
+// previously loaded shared object.
+// An event will be created for each SO it has collisions with.
+// The main efficiency features it uses is caching examined SOs symbols and collision check results.
+// The hope is that after few processes are executed, all major libraries in the system will be cached already
+// and most libraries combinations collisions check results will be cached too.
+func (soColGen *SymbolsCollisionArgsGenerator) handleSOLoaded(event trace.Event) ([][]interface{}, []error) {
+	loadingObjectInfo, err := getSharedObjectInfo(event)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	processLoadedObjects, ok := soColGen.processesSOsCache.GetProcessLoadedObjects(event.HostProcessID)
+	if !ok {
+		processLoadedObjects = []sharedobjs.ObjInfo{}
+	}
+	newProcessLoadedObjects := append(processLoadedObjects, loadingObjectInfo)
+	soColGen.processesSOsCache.SetProcessLoadedObjects(event.HostProcessID, newProcessLoadedObjects)
+
+	// Exported symbols will be updated if needed in the findSOCollisions method
+	loadingObject := loadingSOInstance{ObjInfo: loadingObjectInfo}
+
+	var collisionEventsArgs [][]interface{}
+	var errs []error
+	for _, lsoInfo := range processLoadedObjects {
+		collisions, err := soColGen.findSOCollisions(&loadingObject, lsoInfo)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if len(collisions) > 0 {
+			collisionEventsArgs = append(collisionEventsArgs, []interface{}{loadingObjectInfo.Path, lsoInfo.Path, collisions})
+		}
+	}
+
+	return collisionEventsArgs, errs
+}
+
+// handleExec delete saved process loaded objects in case of execve, because entire memory is overwritten in this case
+func (soColGen *SymbolsCollisionArgsGenerator) handleExec(event trace.Event) ([][]interface{}, []error) {
+	soColGen.processesSOsCache.SetProcessLoadedObjects(event.HostProcessID, []sharedobjs.ObjInfo{})
+	return nil, nil
+}
+
+// findSOCollisions check for collisions between new SO loaded to existing one in the most efficient way possible.
+// It updates the given SO if symbols are missing.
+func (soColGen *SymbolsCollisionArgsGenerator) findSOCollisions(so *loadingSOInstance, loadedSO sharedobjs.ObjInfo) ([]string, error) {
+	var err error
+	collisions, ok := soColGen.collisionsCache.GetCollision(so.Id, loadedSO.Id)
+	if !ok {
+		if so.ExportedSymbols == nil {
+			so.ExportedSymbols, err = soColGen.soLoader.GetExportedSymbols(so.ObjInfo)
+			if err != nil {
+				// TODO: rate limit frequent errors for overloaded envs
+				_, ok := soColGen.returnedErrors[err.Error()]
+				if !ok {
+					soColGen.returnedErrors[err.Error()] = true
+					logger.Warn("symbols_loaded", "object loaded", so.ObjInfo, "error", err.Error())
+				} else {
+					logger.Debug("symbols_loaded", "object loaded", so.ObjInfo, "error", err.Error())
+				}
+				return nil, nil
+			}
+			so.FilterSymbols(soColGen.symbolsBlackList)
+			so.FilterOutSymbols(soColGen.symbolsWhiteList)
+			if err != nil {
```

Comment:
It would already have returned here. I'm changing (in a new PR where I'm adding more documentation and removing the initialization from the initDerivationTable function) this. Removing this err check and returning `nil, err` in the check above.